File: compiler.py
# Token types
INTEGER = 'INTEGER'
PLUS = 'PLUS'
MINUS = 'MINUS'
MULTIPLY = 'MULTIPLY'
DIVIDE = 'DIVIDE'
EOF = 'EOF'

# Token class
class Token:
    def __init__(self, type, value):
        self.type = type
        self.value = value

    def __str__(self):
        return f'Token({self.type}, {self.value})'

    def __repr__(self):
        return self.__str__()

# Lexer
class Lexer:
    def __init__(self, text):
        self.text = text
        self.pos = 0
        self.current_char = self.text[self.pos]

    def error(self):
        raise Exception('Invalid character')

    def advance(self):
        self.pos += 1
        if self.pos < len(self.text):
            self.current_char = self.text[self.pos]
        else:
            self.current_char = None

    def skip_whitespace(self):
        while self.current_char is not None and self.current_char.isspace():
            self.advance()

    def integer(self):
        result = ''
        while self.current_char is not None and self.current_char.isdigit():
            result += self.current_char
            self.advance()
        return int(result)

    def get_next_token(self):
        while self.current_char is not None:
            if self.current_char.isspace():
                self.skip_whitespace()
                continue

            if self.current_char.isdigit():
                return Token(INTEGER, self.integer())

            if self.current_char == '+':
                self.advance()
                return Token(PLUS, '+')

            if self.current_char == '-':
                self.advance()
                return Token(MINUS, '-')

            if self.current_char == '*':
                self.advance()
                return Token(MULTIPLY, '*')

            if self.current_char == '/':
                self.advance()
                return Token(DIVIDE, '/')

            self.error()

        return Token(EOF, None)

# Parser
class Parser:
    def __init__(self, lexer):
        self.lexer = lexer
        self.current_token = self.lexer.get_next_token()

    def error(self):
        raise Exception('Invalid syntax')

    def eat(self, token_type):
        if self.current_token.type == token_type:
            self.current_token = self.lexer.get_next_token()
        else:
            self.error()

    def factor(self):
        token = self.current_token
        self.eat(INTEGER)
        return token.value

    def term(self):
        result = self.factor()

        while self.current_token.type in (MULTIPLY, DIVIDE):
            token = self.current_token
            if token.type == MULTIPLY:
                self.eat(MULTIPLY)
                result *= self.factor()
            elif token.type == DIVIDE:
                self.eat(DIVIDE)
                result /= self.factor()

        return result

    def expr(self):
        result = self.term()

        while self.current_token.type in (PLUS, MINUS):
            token = self.current_token
            if token.type == PLUS:
                self.eat(PLUS)
                result += self.term()
            elif token.type == MINUS:
                self.eat(MINUS)
                result -= self.term()

        return result

# Compiler
def main():
    while True:
        try:
            text = input('> ')
        except EOFError:
            break

        if not text:
            continue

        lexer = Lexer(text)
        parser = Parser(lexer)
        result = parser.expr()
        print(result)

# Run the compiler
if __name__ == '__main__':
    main()

File: compiler.cpp
#include <iostream>
#include <string>
#include <sstream>
#include <cctype>

enum class TokenType {
    INTEGER,
    PLUS,
    MINUS,
    MULTIPLY,
    DIVIDE,
    END_OF_INPUT
};

struct Token {
    TokenType type;
    int value;

    Token(TokenType t, int v) : type(t), value(v) {}
};

class Lexer {
public:
    Lexer(const std::string& input) : input_(input), position_(0) {}

    Token getNextToken() {
        while (position_ < input_.size() && std::isspace(input_[position_])) {
            ++position_;
        }

        if (position_ >= input_.size()) {
            return Token(TokenType::END_OF_INPUT, 0);
        }

        char currentChar = input_[position_];
        if (std::isdigit(currentChar)) {
            return Token(TokenType::INTEGER, parseInt());
        } else {
            ++position_;
            switch (currentChar) {
                case '+':
                    return Token(TokenType::PLUS, 0);
                case '-':
                    return Token(TokenType::MINUS, 0);
                case '*':
                    return Token(TokenType::MULTIPLY, 0);
                case '/':
                    return Token(TokenType::DIVIDE, 0);
                default:
                    throw std::runtime_error("Invalid token");
            }
        }
    }

private:
    int parseInt() {
        std::stringstream ss;
        while (position_ < input_.size() && std::isdigit(input_[position_])) {
            ss << input_[position_++];
        }
        int result;
        ss >> result;
        return result;
    }

    std::string input_;
    size_t position_;
};

class Parser {
public:
    Parser(Lexer& lexer) : lexer_(lexer) {}

    int parseExpression() {
        int left = parseTerm();
        while (true) {
            Token token = lexer_.getNextToken();
            switch (token.type) {
                case TokenType::PLUS:
                    left += parseTerm();
                    break;
                case TokenType::MINUS:
                    left -= parseTerm();
                    break;
                default:
                    return left;
            }
        }
    }

    int parseTerm() {
        int left = parseFactor();
        while (true) {
            Token token = lexer_.getNextToken();
            switch (token.type) {
                case TokenType::MULTIPLY:
                    left *= parseFactor();
                    break;
                case TokenType::DIVIDE:
                    left /= parseFactor();
                    break;
                default:
                    return left;
            }
        }
    }

    int parseFactor() {
        Token token = lexer_.getNextToken();
        if (token.type == TokenType::INTEGER) {
            return token.value;
        } else {
            throw std::runtime_error("Expected integer");
        }
    }

private:
    Lexer& lexer_;
};

int main() {
    std::string input;
    std::cout << "Enter an arithmetic expression: ";
    std::cin >> input;

    Lexer lexer(input);
    Parser parser(lexer);

    try {
        int result = parser.parseExpression();
        std::cout << "Result: " << result << std::endl;
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
    }

    return 0;
}
